pkg/endpoint/bpf.go

```diff
@@ -1129,60 +1097,104 @@ func (e *Endpoint) ApplyPolicyMapChanges(proxyWaitGroup *completion.WaitGroup) e
-	//  from the desired policy here.
-	//  ConsumeMapChanges() applies the incremental updates to the
-	//  desired policy and only returns changes that need to be
-	//  applied to the Endpoint's bpf policy map.
+	// Always update Envoy if policy has changed
+	updateEnvoy := hasNewPolicy
+
+	// Note that after successful endpoint regeneration the desired and realized policies are
+	// the same pointer. During the bpf regeneration possible incremental updates are collected
+	// on the newly computed desired policy, which is not fully realized yet. This is why we get
+	// the map changes from the desired policy here.
+
+	// ConsumeMapChanges() applies the incremental updates to the desired policy and only
+	// returns changes that need to be applied to the Endpoint's bpf policy map.
 	closer, changes := e.desiredPolicy.ConsumeMapChanges()
 	defer closer()
 
-	if changes.Empty() {
-		// no changes, nothing to do
-		return nil
+	hasEnvoyRedirect := e.desiredPolicy.L4Policy.HasEnvoyRedirect()
+	if !changes.Empty() {
+		// updateEnvoy if there were any mapChanges, but only if the endpoint has Envoy
+		// redirects, or is an Ingress endpoint, which needs to enforce also the full L3/4
+		// policy.
+		//
+		// Even if there are no changes, we update the proxyWaitGroup for any in-progress
+		// NetworkPolicy update to be done if the endpoint has envoy redirects, so that the
+		// the expected policy is in place.
+		//
+		// 'updateEnvoy' is already set to 'true' if policy changed. In that case there can
+		// be new redirects and a full policy map update even if there were no incremental
+		// updates.
+		updateEnvoy = updateEnvoy || hasEnvoyRedirect || e.isIngress
 	}
 
+	stats := &regenContext.Stats
+	datapathRegenCtxt := regenContext.datapathRegenerationContext
+	var err error
+
+	proxyWaitGroup := datapathRegenCtxt.proxyWaitGroup
+
 	// Ingress endpoint does not need to wait.
 	// This also lets daemon/cmd integration tests to proceed
 	if e.isProperty(PropertySkipBPFPolicy) {
+		if logging.CanLogAt(log.Logger, logrus.DebugLevel) {
+			log.WithField(logfields.EndpointID, e.ID).Debug("Ingress Endpoint updating Network policy")
+		}
+		proxyWaitGroup = nil
+	}
+
+	// Configure the new network policy with the proxies.
+	//
+	// This must be done after adding new redirects, as waiting for policy update ACKs is
+	// disabled when there are no listeners, which is the case before the first redirect is
+	// added.
+	//
+	// Do this before updating the bpf policy maps below, so that the proxy listeners have a
+	// chance to be ready when new traffic is redirected to them.
+	// NOTE: unlike regeneratePolicy, updateNetworkPolicy requires the endpoint read lock
+	if updateEnvoy {
+		e.getLogger().
+			WithField(logfields.SelectorCacheVersion, e.desiredPolicy.VersionHandle).
+			Debug("applyPolicyMapChanges: Updating Envoy NetworkPolicy")
+		stats.proxyPolicyCalculation.Start()
+		var rf revert.RevertFunc
+		err, rf = e.updateNetworkPolicy(proxyWaitGroup)
```

Comment:
Idea: rename `updateNetworkPolicy()` to `updateProxyPolicy()`? And maybe we can just get rid of `updateNetworkPolicy()` entirely?

Reply:
Further commentary: the DNS proxy doesn't care about incremental updates, so the assumption that "proxy is envoy" is true here. However, the method names do not make this distinction, and the code generally multiplexes Envoy vs. DNS in `pkg/proxy/Proxy` rather than `pkg/endpoint/Endpoint`

Does it make sense to push any of the "should we update Envoy?" predicates in to `pkg/proxy`? It all seems a bit awkward for endpoints to care about Envoy in particular.

Reply:
Got rid of it! Better not give chances calling it anywhere else...